src/Classes/Canvas.ts

```diff
@@ -154,7 +158,159 @@ export class Canvas {
+                r: Math.floor(Utils.mixComponents(sourceColor.r, destColor.r, p1.radius, p2.radius)),
+            };
+        } else {
+            colorLine = container.particles.lineLinkedColor as IRgb;
+        }
+
+        const width = container.retina.lineLinkedWidth;
+
+        CanvasUtils.drawLineLinked(ctx, width, pos1, pos2, options.backgroundMask.enable, colorLine, opacity);
+    }
+
+    public drawConnectLine(p1: Particle, p2: Particle): void {
+        const lineStyle = this.lineStyle(p1, p2);
+
+        if (!lineStyle) {
+            return;
+        }
+
+        const ctx = this.context;
+
+        if (!ctx) {
+            return;
+        }
+
+        CanvasUtils.drawConnectLine(ctx, this.container.retina.lineLinkedWidth, lineStyle, p1.position, p2.position);
+    }
+
+    public drawGrabLine(particle: Particle, opacity: number, mousePos: ICoordinates): void {
+        const container = this.container;
+        const options = container.options;
+        const optColor = options.particles.lineLinked.color;
+
+        let lineColor = container.particles.lineLinkedColor || Utils.hexToRgb(optColor);
+
+        if (lineColor == "random") {
+            lineColor = Utils.getRandomColorRGBA();
+        }
+
+        container.particles.lineLinkedColor = lineColor;
+
+        let colorLine: IRgb = {r: 127, g: 127, b: 127};
+        const ctx = container.canvas.context;
+
+        if (!ctx) {
+            return;
+        }
+
+        if (container.particles.lineLinkedColor == "random") {
+            colorLine = Utils.getRandomColorRGBA();
+        } else {
+            colorLine = container.particles.lineLinkedColor as IRgb || colorLine;
+        }
+
+        const beginPos = {
+            x: particle.position.x + particle.offset.x,
+            y: particle.position.y + particle.offset.y,
+        };
+
+        CanvasUtils.drawGrabLine(ctx, container.retina.lineLinkedWidth, beginPos, mousePos, colorLine, opacity);
+    }
+
+    public drawParticle(particle: Particle): void {
+        const container = this.container;
+        const options = container.options;
+
+        let radius: number;
+        let opacity: number;
+        let colorValue: string | undefined;
+
+        if (particle.bubbler.radius !== undefined) {
+            radius = particle.bubbler.radius;
+        } else {
+            radius = particle.radius;
+        }
+
+        if (particle.bubbler.opacity !== undefined) {
+            opacity = particle.bubbler.opacity;
+        } else {
+            opacity = particle.opacity.value;
+        }
+
+        if (particle.color) {
+            colorValue = `rgba(${particle.color.r}, ${particle.color.g}, ${particle.color.b}, ${opacity})`;
+        }
+
+        if (!this.context || !colorValue) {
+            return;
+        }
+
+        CanvasUtils.drawParticle(this.context, particle, colorValue, options.backgroundMask.enable, radius, options.particles.shape.stroke);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Exceeds maximum line length of 120](https://app.codacy.com/manual/ar3s/tsparticles/pullRequest?prid=5171185)